### ISSUE TITLE
[Bug] `model.evaluate()` reports incorrect metric values in graph mode when dataset has fixed batch dimension

### ISSUE THREAD

**Issue by stinodego** [keras-team/keras#23084]
When calling `model.evaluate()` with `run_eagerly=False` on a dataset whose `element_spec` has a **fixed batch dimension** (e.g. from `tf.data.Dataset.from_generator` with a pre-batched `output_signature`), stateful metrics report diluted/incorrect values.

The same model + metric reports the correct value when:
- `run_eagerly=True`, or
- the dataset uses a `None` batch dimension (e.g. `from_tensor_slices().batch()`)

The dilution scales with `batch_size / n_total_samples`, consistent with one phantom batch being added to the metric accumulator during `tf.function` tracing.

## Minimal reproduction

```python
import keras
import numpy as np
import tensorflow as tf

def make_prebatched_dataset(n_batches=2, batch_size=128, seq_len=10):
    """Pre-batched dataset (batch dim in signature) — triggers the bug."""

    def gen():
        rng = np.random.default_rng(42)
        for _ in range(n_batches):
            x = rng.random((batch_size, seq_len)).astype(np.float32)
            y = rng.random((batch_size, seq_len)).astype(np.float32)
            yield x, y

    sig = (
        tf.TensorSpec((batch_size, seq_len), tf.float32),
        tf.TensorSpec((batch_size, seq_len), tf.float32),
    )
    return tf.data.Dataset.from_generator(gen, output_signature=sig)

def create_model(run_eagerly):
    inputs = keras.Input(shape=(10,))
    outputs = keras.layers.Dense(10, kernel_initializer=keras.initializers.GlorotUniform(seed=0))(inputs)
    model = keras.Model(inputs, outputs)
    model.compile(
        loss="mse",
        metrics=[keras.metrics.MeanAbsoluteError(name="mae")],
        run_eagerly=run_eagerly,
    )
    return model

# Graph mode — BUG: metric was diluted
model_graph = create_model(run_eagerly=False)
graph = model_graph.evaluate(make_prebatched_dataset(), return_dict=True, verbose=0)

# Eager mode — correct
model_eager = create_model(run_eagerly=True)
eager = model_eager.evaluate(make_prebatched_dataset(), return_dict=True, verbose=0)

print(f"graph mae: {graph['mae']:.6f}")
print(f"eager mae: {eager['mae']:.6f}")
print(f"gap:       {eager['mae'] - graph['mae']:.6f}")
assert abs(eager["mae"] - graph["mae"]) < 0.001, "BUG: graph != eager"
```

## Expected behavior
`model.evaluate()` should return the same metric value regardless of `run_eagerly`.

## Actual behavior
Graph mode reports a lower metric value. The gap increases with larger `batch_size` relative to total samples, suggesting one phantom batch is added to the metric accumulator during `tf.function` compilation when the input spec has a fully-defined (non-`None`) batch dimension.

## Workaround
Force eager execution during evaluate:

```python
model.run_eagerly = True
model.test_function = None
model.evaluate(...)
```

Or introduce a None batch dimension by applying `.unbatch().batch(batch_size)` to the dataset before evaluation.

## Environment

Python 3.12
TensorFlow 2.19.0
Keras 3.14.1
Backend: TensorFlow
OS: macOS (also confirmed on Linux/GPU via SageMaker)

**Comment by dhantule** [comment]
Hi @stinodego, thanks for reporting this. 

I have reproduced this issue with Keras 3.14.1 in this [gist](https://colab.sandbox.google.com/gist/dhantule/b53b800b68566bce3bb6882e733abb55/untitled208.ipynb#scrollTo=DcvH_mFUBt3Z), we'll look into this and update you.

**Comment by stinodego** [comment]
I made a small update the the reproducible example: I did not seed the model creation so the scores would differ slightly even without the bug. The repro is now correct.

I also coded a solution. It's a small fix in the TensorFlow trainer. See: https://github.com/stinodego/keras/pull/1

I am leaving on a holiday for 10 days, so I will wait with opening a PR until I'm back as I cannot respond to feedback in the coming days.
If a maintainer wants to solve this issue or take my solution and merge it before then, that is completely fine with me. I just want the issue fixed as it is affecting my team.

**Comment by maitry63** [comment]
Hi @stinodego , Thank you for your detailed input!

However, you created a PR in your personal repo, you should create in the official keras repo.
You can go through this [Documentation](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md).

Also, created a corresponding PR for the fix.

**Comment by stinodego** [comment]
Hey @maitry63 ! I was aware my PR was on my personal repo - I didn't expect you to react so quickly :) I will be leaving on vacation soon so I wanted to avoid the situation that you respond to my PR and I won't be there to react. Great that you have picked this up!

I found that with this fix, it seems that in `multi_step_on_iterator` we no longer need the special case for `if self.steps_per_execution == 1:`. So with this PR you could also remove that dead code (if I am understanding this correctly).

In any case, thanks a lot for picking this up!

**Comment by maitry63** [comment]
Hi @stinodego , Thank you for the clarification  &  the suggestion.

I think we should keep the `if self.steps_per_execution == 1: `branch inside `multi_step_on_iterator()`.

Even though the single-step case is now handled in the outer `function()` fast path, `multi_step_on_iterator()` can still be called directly from other execution paths.

Because of that, the check inside `multi_step_on_iterator()` acts as a safety fallback to ensure correct behaviour even when it is used independently of the wrapper. Removing it could unintentionally reintroduce the original tracing-related issue in edge cases.

So I would prefer to keep it for correctness and defensive coverage.

Have a nice vacation!

**Comment by stinodego** [comment]
> `multi_step_on_iterator()` can still be called directly from other execution paths.

Ah, I must have missed this. Thanks for the correction and the PR looks good to me as is! Very happy with the quick fix 👌

**Comment by stinodego** [comment]
Hey @maitry63 - sorry to ping you but I wonder if there's anything I can do to move this forward? Or should we just wait for a maintainer that can review your fix?

**Comment by maitry63** [comment]
Hi @stinodego , Sorry for the delay! We can wait for the maintainer’s review. We’ll try to get this checked on priority. 
Thank you.

**Comment by stinodego** [comment]
Fixed by https://github.com/keras-team/keras/pull/23179

Thanks so much for your help!
